### PR TITLE
improvement(nemesis.py): refactor nodetool_cleanup to run in parallel

### PR DESCRIPTION
**Pull request by Mark-Gurevich**
Based on task: https://github.com/scylladb/qa-tasks/issues/884 Refactoring nodetool_cleanup to be called by threadPoolExecutor (parallel).

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/nemesis.py [hunk 1]
```diff
@@ -1732,10 +1732,13 @@ def repair_nodetool_rebuild(self):
     def disrupt_nodetool_cleanup(self):
         # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
         test_keyspaces = self.cluster.get_test_keyspaces()
-        for node in self.cluster.nodes:
+
+        def clean_task(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             for keyspace in test_keyspaces:
                 node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+        with ThreadPoolExecutor() as executor:
```
Comment (fruch) [inline sdcm/nemesis.py:1740]: How many executers we are using here ?

File: sdcm/nemesis.py [hunk 2]
```diff
@@ -1732,10 +1732,13 @@ def repair_nodetool_rebuild(self):
...
+            executor.map(clean_task, self.cluster.nodes)
```
Comment (fruch) [inline sdcm/nemesis.py:1741]: Where do we wait for those to finish ? where do we get failures from those calls ?

**Review by fruch — CHANGES_REQUESTED** [review]
We should be more specific on how many threads whould run in parallel and sure we wait for results.

Look at the ParallelObject, and consider using it directly.

File: sdcm/nemesis.py (hunk 1 above)
Comment (vponomaryov) [reply sdcm/nemesis.py:1740]: According to the source code it is as of following: `max_workers = min(32, (os.cpu_count() or 1) + 4)`
Or from `5` to `32` if not specified.

File: sdcm/nemesis.py (hunk 2 above)
Comment (vponomaryov) [reply sdcm/nemesis.py:1741]: Agree, for now it is `fire and run`. We must wait for the result and check it.

**Comment by soyacz** [comment]
@Mark-Gurevich do you have some example run?

File: sdcm/nemesis.py [hunk 3]
```diff
@@ -1830,14 +1830,24 @@ def repair_nodetool_rebuild(self):
...
+            test_keyspaces = self.cluster.get_test_keyspaces()
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
                 for keyspace in test_keyspaces:
                     node.run_nodetool(sub_cmd="cleanup", args=keyspace)
 
+        results = ParallelObject(self.cluster.nodes, num_workers=min(32, len(self.cluster.nodes))).\
+            run(_dis_nodetool_cleanup, ignore_exceptions=True, unpack_objects=True)
+        runs_that_finished_with_exception = [res for res in results if res.exc]
+        if runs_that_finished_with_exception:
```
Comment (vponomaryov) [inline sdcm/nemesis.py:1845]: Use generator to avoid processing of redundant elements:
```
if any(thread_result for thread_result in thread_results if thread_result.exc):
    ...
else:
    ...
```

File: sdcm/nemesis.py [hunk 4]
```diff
@@ -1830,14 +1830,24 @@ def repair_nodetool_rebuild(self):
...
+            self.log.error("Nodetool cleanup failed on one of the nodes",
```
Comment (vponomaryov) [inline sdcm/nemesis.py:1846]: The number of failures is unknown. So, need to rephrase it.

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/nemesis.py [hunk 5]
```diff
@@ -1830,14 +1830,24 @@ def repair_nodetool_rebuild(self):
             self.target_node.run_nodetool('rebuild', retry=0)
 
     def disrupt_nodetool_cleanup(self):
-        # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
-        test_keyspaces = self.cluster.get_test_keyspaces()
-        for node in self.cluster.nodes:
+        # Inner disrupt function for ParallelObject
+        def _dis_nodetool_cleanup(node):
```
[the remark this answers is not in the document]
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1834]: We agreed with Roy to leave it run in parallel only on the nodes. No need to run in parallel on keyspaces.

File: sdcm/nemesis.py (hunk 3 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1845]: Changed.

The same code is present in common.py lines 436-439.
`runs_that_finished_with_exception = [res for res in results if res.exc]
        if runs_that_finished_with_exception:
            raise ParallelObjectException(results=results)
        return results`
Should we change it there as well?

File: sdcm/nemesis.py [hunk 6]
```diff
@@ -1830,14 +1830,24 @@ def repair_nodetool_rebuild(self):
...
+            # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
+            test_keyspaces = self.cluster.get_test_keyspaces()
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
                 for keyspace in test_keyspaces:
                     node.run_nodetool(sub_cmd="cleanup", args=keyspace)
 
+        results = ParallelObject(self.cluster.nodes, num_workers=min(32, len(self.cluster.nodes))).\
```
[the remark this answers is not in the document]
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1842]: Done

File: sdcm/nemesis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vponomaryov) [reply sdcm/nemesis.py:1834]: Ok.

File: sdcm/nemesis.py (hunk 4 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1846]: Even though we don't have number of failures the error message will include all the exceptions.
If you say its important to print the exact number of failures in this text message we should not use generator in previous code block. The length of list 'runs_that_finished_with_exception' gives us exactly what you ask for.

**Comment by Mark-Gurevich** [comment]
> @Mark-Gurevich do you have some example run?

@soyacz I tested this code locally on docker (happy path and bad path by changing the command to a wrong one). 
How would you suggest to check 'bad path' on jenkins?

File: sdcm/nemesis.py (hunk 3 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: ok

File: sdcm/nemesis.py (hunk 4 above)
Comment (vponomaryov) [reply sdcm/nemesis.py]: I don't ask for number. `The number of failures is unknown. So, need to rephrase it.` means `the log message's statement is not guaranteed and need to rephrase it to something that will be true in any case`.

File: sdcm/nemesis.py [hunk 7]
```diff
@@ -1830,14 +1830,22 @@ def repair_nodetool_rebuild(self):
             self.target_node.run_nodetool('rebuild', retry=0)
 
     def disrupt_nodetool_cleanup(self):
-        # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
-        test_keyspaces = self.cluster.get_test_keyspaces()
-        for node in self.cluster.nodes:
+        # Inner disrupt function for ParallelObject
+        def _dis_nodetool_cleanup(node):
+            # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
+            test_keyspaces = self.cluster.get_test_keyspaces()
```
[the remark this answers is not in the document]
Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: You are totally right. I should move this up from the inner function.
Done

File: sdcm/nemesis.py [hunk 8]
```diff
@@ -1830,14 +1830,23 @@ def repair_nodetool_rebuild(self):
             self.target_node.run_nodetool('rebuild', retry=0)
 
     def disrupt_nodetool_cleanup(self):
-        # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
         test_keyspaces = self.cluster.get_test_keyspaces()
-        for node in self.cluster.nodes:
+
+        # Inner disrupt function for ParallelObject
+        def _dis_nodetool_cleanup(node):
+            # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
```
Comment (vponomaryov) [inline sdcm/nemesis.py:1837]: This comment is related to the `test_keyspaces = self.cluster.get_test_keyspaces()` line.
So, need to return it back too.

Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: My bad... Done.

**Review by fgelcer — APPROVED** [review]
LGTM

**Comment by fgelcer** [comment]
@Mark-Gurevich said that the staging run failed... please add to here at least the link to the job, so we can from there start discussion on the problem, and help to investigate the failure

**Comment by Mark-Gurevich** [comment]
> @Mark-Gurevich said that the staging run failed... please add to here at least the link to the job, so we can from there start discussion on the problem, and help to investigate the failure

The job failed after 1 hour and 43 minutes.
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/mark/job/longevity-5gb-1h-NodeToolCleanupMonkey-aws-test-parallelObject/

```
18:08:00  ----- LAST ERROR EVENT -------------------------------------------------------
18:08:00  2023-05-03 15:00:10.557: (NodetoolEvent Severity.ERROR) period_type=end event_id=83c999e4-0a9e-45b8-9a6d-f0e97013cacc duration=3m4s: nodetool_command=cleanup node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3 errors=["Encountered a bad command exit code!\n\nCommand: '/usr/bin/nodetool  cleanup keyspace1'\n\nExit code: 2\n\nStdout:\n\n\n\nStderr:\n\n\tat sun.rmi.transport.Transport.serviceCall(Transport.java:196)\n\tat sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:750)\n\n\n", 'Traceback (most recent call last):\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2554, in run_nodetool\n    self.remoter.run(cmd, timeout=timeout, ignore_status=ignore_status, verbose=verbose, retry=retry)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run\n    result = _run()\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner\n    return func(*args, **kwargs)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run\n    return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute\n    result = connection.run(**command_kwargs)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 620, in run\n    return self._complete_run(channel, exception, timeout_reached, timeout, result, warn, stdout, stderr)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 654, in _complete_run\n    raise UnexpectedExit(result)\nsdcm.remote.libssh2_client.exceptions.UnexpectedExit: Encountered a bad command exit code!\n\nCommand: \'/usr/bin/nodetool  cleanup keyspace1\'\n\nExit code: 2\n\nStdout:\n\n\n\nStderr:\n\n\tat sun.rmi.transport.Transport.serviceCall(Transport.java:196)\n\tat sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:750)\n\n\n\n']
18:08:00  Traceback (most recent call last):
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2554, in run_nodetool
18:08:00  self.remoter.run(cmd, timeout=timeout, ignore_status=ignore_status, verbose=verbose, retry=retry)
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run
18:08:00  result = _run()
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner
18:08:00  return func(*args, **kwargs)
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run
18:08:00  return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute
18:08:00  result = connection.run(**command_kwargs)
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 620, in run
18:08:00  return self._complete_run(channel, exception, timeout_reached, timeout, result, warn, stdout, stderr)
18:08:00  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 654, in _complete_run
18:08:00  raise UnexpectedExit(result)
18:08:00  sdcm.remote.libssh2_client.exceptions.UnexpectedExit: Encountered a bad command exit code!
18:08:00  Command: '/usr/bin/nodetool  cleanup keyspace1'
18:08:00  Exit code: 2
18:08:00  Stdout:
18:08:00  Stderr:
18:08:00  at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
18:08:00  at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
18:08:00  at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
18:08:00  at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
18:08:00  at java.security.AccessController.doPrivileged(Native Method)
18:08:00  at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
18:08:00  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
18:08:00  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
18:08:00  at java.lang.Thread.run(Thread.java:750)
18:08:00  ----- LAST WARNING EVENT -----------------------------------------------------
18:08:00  2023-05-03 13:34:49.055 <2023-05-03 13:34:49.000>: (DatabaseLogEvent Severity.WARNING) period_type=one-time event_id=e0becb4a-46d5-4c04-9470-eec99ab65464: type=RPC_CONNECTION regex=(^ERROR|!ERR).*rpc - client .*(connection dropped|fail to connect) line_number=838 node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3
18:08:00  2023-05-03T13:34:49+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3      !ERR | scylla[5513]:  [shard 0] rpc - client 10.4.0.84:7001: client connection dropped: recv: Connection reset by peer
18:08:00  ----- LAST NORMAL EVENT ------------------------------------------------------
18:08:00  2023-05-03 14:59:56.818: (NodetoolEvent Severity.NORMAL) period_type=end event_id=3299dff0-714b-413b-b424-0b8993722002 duration=2m26s: nodetool_command=cleanup node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-1
```

**Comment by soyacz** [comment]
@Mark-Gurevich what scylladb logs show?

**Comment by Mark-Gurevich** [comment]
node 3:
`2023-05-03 14:57:33.811 <2023-05-03 14:57:33.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=54fc672b-8572-464b-b1b3-0ac4b76b30f6: type=RUNTIME_ERROR regex=std::runtime_error line_number=2684 node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3
2023-05-03T14:57:33+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3      !ERR | scylla[5513]:  [shard 0] api - Failed force_keyspace_cleanup of keyspace1.table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}: std::runtime_error (cleanup request failed: there is an ongoing cleanup on keyspace1.standard1)`

node 2:
`2023-05-03 14:58:05.706 <2023-05-03 14:58:05.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=54fc672b-8572-464b-b1b3-0ac4b76b30f6: type=RUNTIME_ERROR regex=std::runtime_error line_number=2856 node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-2
2023-05-03T14:58:05+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-2      !ERR | scylla[5376]:  [shard 0] api - Failed force_keyspace_cleanup of keyspace1.table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}: std::runtime_error (cleanup request failed: there is an ongoing cleanup on keyspace1.standard1)`

**Comment by fruch** [comment]
> > I remembered there is a limitation like this from the past, but I couldn't reproduce it. What was the test doing? did it try to run a cleanup on the same keyspace from different nodes at the same time?
> 
> it being part of the parallel `nodetool cleanup` runs, the nemesis is now running cleanup on all nodes in parallel, without specifying keyspaces... if that is a limitation, then what needs to be done, is to list the keyspaces available, and run one on each node, using kid of round robin (so if we have 2 keyspaces and 4 nodes, only the 2 first nodes will run it... or if we have 8 keyspaces and 4 nodes, each will run 2)...

I wouldn't go there, cleanup is supposed to be local and clean data doesn't own, it shouldn't be blocked by other nodes.

I would raise it with scylla core first, and if it isn't supported for a good reason, we shouldn't be doing it.

**Comment by fruch** [comment]
> > I remembered there is a limitation like this from the past, but I couldn't reproduce it. What was the test doing? did it try to run a cleanup on the same keyspace from different nodes at the same time?
> 
> it being part of the parallel `nodetool cleanup` runs, the nemesis is now running cleanup on all nodes in parallel, without specifying keyspaces... if that is a limitation, then what needs to be done, is to list the keyspaces available, and run one on each node, using kid of round robin (so if we have 2 keyspaces and 4 nodes, only the 2 first nodes will run it... or if we have 8 keyspaces and 4 nodes, each will run 2)...

I wouldn't go there, cleanup is supposed to be local and clean data doesn't own, it shouldn't be blocked by other nodes.

I would raise it with scylla core first, and if it isn't supported for a good reason, we shouldn't be doing it.

**Comment by roydahan** [comment]
Yes, we first need to understand the limitation.
When I created this task I remembered this limitation but couldn't find anywhere it even in scylla code.
Let's raise an issue in scylla to better understand what exactly the limitation is (include the message we've got of course and explain what exactly you did).

**Comment by Mark-Gurevich** [comment]
I've opened the issue with error messages, nodetool error events.
https://github.com/scylladb/scylladb/issues/13828

File: sdcm/nemesis.py [hunk 9]
```diff
@@ -1831,14 +1831,14 @@ def repair_nodetool_rebuild(self):
...
+        # test_keyspaces = self.cluster.get_test_keyspaces()
 
         # Inner disrupt function for ParallelObject
         def _dis_nodetool_cleanup(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
-                for keyspace in test_keyspaces:
-                    node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+                # for keyspace in test_keyspaces:
+                node.run_nodetool(sub_cmd="cleanup")
```
Comment (roydahan) [inline sdcm/nemesis.py:1841]: So where is the parallelism here?
It's supposed to be "for node in db_nodes:"

**Comment by Mark-Gurevich** [comment]
The parallelism is done by parallelObject.

threads = ParallelObject(self.cluster.nodes, num_workers=min(32,
len(self.cluster.nodes)))
thread_results = threads.run(_dis_nodetool_cleanup,
ignore_exceptions=True, unpack_objects=True)

File: sdcm/nemesis.py (hunk 9 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1841]: The parallelism is done by parallelObject.

threads = ParallelObject(self.cluster.nodes, num_workers=min(32,
len(self.cluster.nodes)))
thread_results = threads.run(_dis_nodetool_cleanup,
ignore_exceptions=True, unpack_objects=True)

Comment (roydahan) [reply sdcm/nemesis.py:1841]: Somehow the lines below didn't show up when I looked at it yesterday (or I saw only the last commit)

Comment (roydahan) [reply sdcm/nemesis.py:1841]: Does it work?
Do we see the cleanup running in parallel on all nodes?

Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1841]: Yes It worked, but for some reason it also tried to run cleanup several times on the same node.
This was unexpected behaviour.
Following the code it should run in parallel on all nodes.
Last commit removed looping over keyspaces, so now we should have one cleanup command per node.
Testing in docker seems to fail for other reason.
I'm launching another test run in aws.

Comment (fgelcer) [reply sdcm/nemesis.py:1841]: we see all the 3 nodes having a call to start `nodetool cleanup` in parallel.

the problem I see is that, we are not waiting for the cleanup to end, before "ending the nemesis" and we eventually started the next nemesis before the previous cleanup ended, causing an exception

File: sdcm/nemesis.py [hunk 10]
```diff
@@ -1831,12 +1831,21 @@ def repair_nodetool_rebuild(self):
...
+        def _dis_nodetool_cleanup(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
-                for keyspace in test_keyspaces:
-                    node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+                # for keyspace in test_keyspaces:
+                node.run_nodetool(sub_cmd="cleanup")
+
+        threads = ParallelObject(self.cluster.nodes, num_workers=min(32, len(self.cluster.nodes)))
+        thread_results = threads.run(_dis_nodetool_cleanup, ignore_exceptions=True, unpack_objects=True)
```
Comment (fruch) [inline sdcm/nemesis.py:1844]: Why are we ignoring exceptions ?

And then not failing the nemesis when finding errors ?

We shouldn't be ignoring failures, we should let them bubble up and fail the nemesis.

File: sdcm/nemesis.py [hunk 11]
```diff
@@ -1831,12 +1831,21 @@ def repair_nodetool_rebuild(self):
...
+        if any(res for res in thread_results if res.exc):
+            self.log.error("Nodetool cleanup has failed", exc_info=ParallelObjectException(results=thread_results))
```
Comment (fruch) [inline sdcm/nemesis.py:1846]: no point logging error like this, no one gonna notice it, error should always strive to raise exceptions

File: sdcm/nemesis.py (hunk 10 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1844]: OK, then I change it to default value ignore_exceptions=False

File: sdcm/nemesis.py (hunk 11 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1846]: I remove this code. If an exception occurs it will be raised by parallelObject.run() inner logic:

```
runs_that_finished_with_exception = [res for res in results if res.exc]
        if runs_that_finished_with_exception:
            raise ParallelObjectException(results=results)
```

File: sdcm/nemesis.py (hunk 10 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: However, I'm not sure if the thread_results needed and what to check in it as if anything goes wrong exception will be raised anyway.

**Comment by Mark-Gurevich** [comment]
> we see all the 3 nodes having a call to start `nodetool cleanup` in parallel.
> 
> the problem I see is that, we are not waiting for the cleanup to end, before "ending the nemesis" and we eventually started the >next nemesis before the previous cleanup ended, causing an exception

@fgelcer @roydahan 

I tried to look into the `ParallelObject` code and it has a `shutdown()` method in the cleanup function inside the `run()` method.
```
    def clean_up(self, futures):
        # if there are futures that didn't run  we cancel them
        for future, _ in futures:
            future.cancel()
        self._thread_pool.shutdown(wait=False)
        # we need to unregister internal function that waits for all threads to finish when interpreter exits
        atexit.unregister(_python_exit)
```

It is called with `wait=False` and from python docs it appears that it will return immediately
https://docs.python.org/3/library/concurrent.futures.html#executor-objects

If wait is True then this method will not return until all the pending futures are done executing and the resources associated with the executor have been freed. **If wait is False then this method will return immediately and the resources associated with the executor will be freed when all pending futures are done executing**

You can avoid having to call this method explicitly if you use the [with]statement, which will shutdown the [Executor] (waiting as if Executor.shutdown() were called with wait set to True):

AFAIU this is reason why we don't wait for the nemesis to finish when using `paralleObject`. It continues main thread execution without waiting for for Executor to finish.

File: sdcm/nemesis.py [hunk 12]
```diff
@@ -1841,12 +1842,18 @@ def repair_nodetool_rebuild(self):
...
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
-                for keyspace in test_keyspaces:
-                    node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+                # for keyspace in test_keyspaces:
+                node.run_nodetool(sub_cmd="cleanup")
+
+        threads = ParallelObject(self.cluster.nodes, num_workers=min(32, len(self.cluster.nodes)), timeout=43200)
+        thread_results = threads.run(_dis_nodetool_cleanup, unpack_objects=True)
+        wait_for_futures(thread_results)
```
Comment (fruch) [inline sdcm/nemesis.py:1856]: ParallelObject is waiting for the futures, I'm not sure why it's needed to wait for them again ?

Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1856]: It does not wait as it appears from here:
file:///home/mark/logs/NodeToolCleanupMonkey-ParallelObject%235/parallel-timelines-report-4ce134c7/longevity-5gb-1h-NodeToolCleanupMon-4ce134c7-parallel-timelines-report.html

This `wait()` also doesn't work as ParallelObjectResult has no required attributes. It does not inherit from `Future` class in concurrent.futures lib.

Comment (fruch) [reply sdcm/nemesis.py:1856]: as I said ParallelObject should be do the waiting on it's own

how come you get to situation it returns without all commands finished its not clear to me, it mean you are not passing a timeout to the nodetool command, also disabling the retries is recommended.

Comment (fruch) [reply sdcm/nemesis.py:1856]: just remove `wait_for_futures(thread_results)` and I think it would work o.k., now that you have 12h timeout

File: sdcm/nemesis.py [hunk 13]
```diff
@@ -1841,12 +1842,18 @@ def repair_nodetool_rebuild(self):
 
     def disrupt_nodetool_cleanup(self):
         # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
-        test_keyspaces = self.cluster.get_test_keyspaces()
-        for node in self.cluster.nodes:
+        # test_keyspaces = self.cluster.get_test_keyspaces()
```
Comment (fruch) [inline sdcm/nemesis.py:1845]: remove the comment

File: sdcm/nemesis.py [hunk 14]
```diff
@@ -1841,12 +1842,18 @@ def repair_nodetool_rebuild(self):
...
+
+        # Inner disrupt function for ParallelObject
+        def _dis_nodetool_cleanup(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
-                for keyspace in test_keyspaces:
-                    node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+                # for keyspace in test_keyspaces:
```
Comment (fruch) [inline sdcm/nemesis.py:1851]: remove the comment

File: sdcm/nemesis.py [hunk 15]
```diff
@@ -1841,12 +1842,18 @@ def repair_nodetool_rebuild(self):
...
     def disrupt_nodetool_cleanup(self):
         # This fix important when just user profile is run in the test and "keyspace1" doesn't exist.
-        test_keyspaces = self.cluster.get_test_keyspaces()
-        for node in self.cluster.nodes:
+        # test_keyspaces = self.cluster.get_test_keyspaces()
+
+        # Inner disrupt function for ParallelObject
+        def _dis_nodetool_cleanup(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
```
Comment (fruch) [inline sdcm/nemesis.py:1850]: kind of contradict the 12h timeout you set for cleanup on all nodes

Comment (Mark-Gurevich) [reply sdcm/nemesis.py:1846]: Changed for the same value in `threads`

File: sdcm/nemesis.py (hunk 14 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: Done

File: sdcm/nemesis.py (hunk 13 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: Done

File: sdcm/nemesis.py (hunk 12 above)
Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: ok

**Comment by Mark-Gurevich** [comment]
Re-trying:
https://argus.scylladb.com/workspace?state=WyI4NTcxZTIyYi04YzRiLTRiZTEtYmU3OS0xZTEwOTBiNWM4MmIiXQ

**Comment by Mark-Gurevich** [comment]
> Re-trying: https://argus.scylladb.com/workspace?state=WyI4NTcxZTIyYi04YzRiLTRiZTEtYmU3OS0xZTEwOTBiNWM4MmIiXQ

Passed.

File: sdcm/nemesis.py [hunk 16]
```diff
@@ -1840,13 +1840,15 @@ def repair_nodetool_rebuild(self):
...
-        for node in self.cluster.nodes:
+        # Inner disrupt function for ParallelObject
+        def _dis_nodetool_cleanup(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
-                for keyspace in test_keyspaces:
-                    node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+                node.run_nodetool(sub_cmd="cleanup")
+
+        threads = ParallelObject(self.cluster.nodes, num_workers=min(
```
Comment (soyacz) [inline sdcm/nemesis.py:1849]: ParallelObject is not returning threads - better rename it.

File: sdcm/nemesis.py [hunk 17]
```diff
@@ -1840,13 +1840,15 @@ def repair_nodetool_rebuild(self):
...
+            32, len(self.cluster.nodes)), timeout=HOUR_IN_SEC * 48)
+        threads.run(_dis_nodetool_cleanup, unpack_objects=True)
```
Comment (soyacz) [inline sdcm/nemesis.py:1851]: Why `unpack_objects=True`? Is it needed?

Comment (Mark-Gurevich) [reply sdcm/nemesis.py]: Removed.

**Comment by Mark-Gurevich** [comment]
Rerunning on master:latest with last change:
https://argus.scylladb.com/test/8571e22b-8c4b-4be1-be79-1e1090b5c82b/runs?additionalRuns[]=5b3ec699-168a-4df4-a399-ba750c50e631

**Review by fgelcer — COMMENTED** [review]
other than i comment i just left, LGTM

**Review by soyacz — APPROVED** [review]
LGTM

File: sdcm/nemesis.py [hunk 18]
```diff
@@ -1840,13 +1840,15 @@ def repair_nodetool_rebuild(self):
...
+        # Inner disrupt function for ParallelObject
+        def _nodetool_cleanup(node):
             InfoEvent('NodetoolCleanupMonkey %s' % node).publish()
             with adaptive_timeout(Operations.CLEANUP, node, timeout=HOUR_IN_SEC * 48):
-                for keyspace in test_keyspaces:
-                    node.run_nodetool(sub_cmd="cleanup", args=keyspace)
+                node.run_nodetool(sub_cmd="cleanup")
+
+        parallel_objects = ParallelObject(self.cluster.nodes, num_workers=min(
+            32, len(self.cluster.nodes)), timeout=HOUR_IN_SEC * 48)
```
[the remark this answers is not in the document]
Comment (soyacz) [reply sdcm/nemesis.py:1850]: I don't remember why it was set like this, but probably there was no timeout at all (and still there's no, adaptive timeout generates SoftTimeout error events only) - so was made only for measurements how much it really takes.

**Review by vponomaryov — APPROVED** [review]
LGTM

File: sdcm/nemesis.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/nemesis.py:1850]: @soyacz we should have a follow-up task for all those place we mesured, to put it a formula, or at least put smaller soft timeouts based on the collected information

**Review by fruch — APPROVED** [review]
LGTM
